modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/SortNode.java

```diff
@@ -91,6 +112,12 @@ else if (!inLoop)
 
         rows.add(row);
 
+        if (limit >= 0 && rows.size() > limit) {
+            AtomicInteger i = new AtomicInteger();
+
+            rows.removeIf(r -> i.incrementAndGet() == rows.size());
```

Comment:
Such approach makes the whole ticket useless, since it's O(N) complexity for insertion instead of O(Log(N)).
Let's use `GridBoundedPriorityQueue`

Reply:
I'm afraid, it removes head instead of tail

Reply:
Fixed

Reply:
Fixed